File: 202-happy-number/202-happy-number.java
class Solution {
    
    public boolean isHappy(int n) {
        int slow = getHappyNumber(n);
        int fast = getHappyNumber(getHappyNumber(n));
        
        while (slow != fast) {
            slow = getHappyNumber(slow);
            fast = getHappyNumber(getHappyNumber(fast));
        }
        
        return slow == 1;
    }
    
    public int getHappyNumber(int n) {
        int sum = 0;
        while (n > 0) {
            int num = (n%10);
            sum += num * num;
            n /= 10;
        }
        return sum;
    }
}

File: 202-HappyNumber/202-HappyNumber.cpp
// Last updated: 2026. 7. 16. 오전 11:27:31
class Solution {
public:
    int next(int n) {
        int sum = 0;
        int m = n;
        while (m) {
            sum += (m%10) * (m%10);
            m /= 10;
        }
        return sum;
    }
    
    bool isHappy(int n) {
        int tortoise = n;
        int hare = next(n);
        while (tortoise != hare && hare != 1) {
            tortoise = next(tortoise);
            hare = next(next(hare));
        }
        return hare == 1;
    }
};